ERCS/erc-7836.md

```diff
@@ -0,0 +1,188 @@
+---
+eip: 7836
+title: Wallet Call Preparation API
+description: Adds JSON-RPC methods for requesting prepared EIP-5792 calls.
+author: Lukas Rosario (@lukasrosario), Conner Swenberg (@ilikesymmetry), Adam Hodges (@ajhodges), Paaras Bhandari (@paarasbhandari)
+discussions-to: https://ethereum-magicians.org/t/new-erc-wallet-call-preparation-api/22456
+status: Draft
+type: Standards Track
+category: ERC
+created: 2024-12-06
+requires: 1193, 5792
+---
+
+## Abstract
+
+This proposal defines complementary JSON-RPC methods to [EIP-5792](./eip-5792.md)'s `wallet_sendCalls`. While `wallet_sendCalls` is used for an app to submit `calls` to be signed and submitted in a wallet's interface, the methods in this proposal are for an application to request prepared calls (where "prepared" depends on the wallet implementation) to be signed and submitted by the application itself.
+
+## Motivation
+
+With more recent developments in account abstraction and session keys, there is an increasing need for applications to be able to sign and submit operations without switching to a wallet interface (e.g. browser extension). This can be tricky because different account implementations might have different call data encoding, signature formats, etc. To remedy this, we need a way for apps to know how to submit an operation for any account implementation.
+
+## Specification
+
+Two new JSON-RPC methods are defined. If the user is interacting with an application on the browser, they can be used by sending the requests to the wallet's [EIP-1193](./eip-1193.md) provider. However, this proposal also describes how an application might interact with a "wallet server" for background processes.
+
+### `wallet_prepareCalls`
+
+Accepts an EIP-5792 `wallet_sendCalls` request, and returns the prepared calls according to the account's implementation.
+
+#### Parameters
+
+```typescript!
+type PrepareCallsParams = [{
```

Comment:
We should add app-provided IDs ([as proposed to be added to 5792](https://github.com/ethereum/EIPs/pull/8826/files#diff-7a494500c37dc4b57613207d62874322198a4ced215f4a0457a0af908bb6364eR70)). This can be used by the wallet as an idempotency key. As a specific example, an ERC-4337 wallet could derive the nonce key/prefix from the app-provided ID to enable high-throughput use cases (while avoiding double-spend transactions by using the app-provided ID as a sort of idempotency key/intent ID)